[#9027] improve CSV download using Datastore dump

Engerrs opened the pull request:
Fixes: #9027 

## advanced datastore filters + fast datastore_search pagination

- `datastore_search` and `datastore_delete` `filters` now accept
  range values and nested AND and OR operations
- `datastore_search` now returns a `next_page` value with
  filters that can be used for fast pagination when records
  are returned sorted by their `_id` field
- datastore dump endpoint now uses fast pagination to more
  quickly stream large datasets as CSV, JSON, etc.

Engerrs committed:
[#9027] improve CSV download using Datastore dump

wardi commented on ckanext/datastore/backend/postgres.py:
```diff
@@ -1536,6 +1548,17 @@ def search_data(context: Context, data_dict: dict[str, Any]):
+
+    # Currently apply Keyset Pagination only to CSV format
+    if records_format == u'csv':
+        m_csv_string = records.rstrip('\n').rsplit('\n', 1)
```
This is checking for more than one line in the records?

Engerrs committed:
[#9027] update keyset pagination logic to support basic sorting and conditions

Engerrs commented:
Hi @wardi , thanks for noticing the sorting issue. I updated the logic a bit to take thins into account.

Keyset Pagination sorting is its issue, when there complex sorting from more then 1 sorting element, in such cases requires to modify the "WHERE" condition into tuples, which makes things more complex and I want to avoid this for now, as we will need to re-write the "where_clause" logic. There also an scenario in which the query becomes even more complex, is where the sorting elements have different sorting order (e.g. one asc and the second desc), this can require splitting the query with "OR"s.

Due to the fact the Keyset Pagination initially was added to support quicker pure download for datastore dump, I would leave for now the logic to be able to support at least basic scenarios of sorting (e.g. asc and desc for _id) and conditioning. For all other more complex filters and sorting, it would use "offset", keeping in mind, that after filtering the amount of results becomes much lower.

Engerrs committed:
[#9027] Fix Lint, Tests

wardi commented:
Thanks @Engerrs 

Instead of adding a new `last_id` parameter to the `datastore_search` API let's implement a special case of the [upcoming generic search filters](https://github.com/ckan/ckanext-search/blob/main/doc/filters.md) feature.

e.g. If `datastore_search` parameters contain:
```
{
  "filters": {
    "_id": {"gt": 123},
    …
  },
  …
}
```
Then treat "123" as the `last_id` value.

This way once the new generic search feature is implemented this will stop being a special case and users won't need to update their API usage. Document it as a special case for efficient pagination in `datastore_search` as part of this PR.

It's fine if it doesn't help with paginating other sort orders since we're being clear that we only want `_id` values greater than the number passed.

In this PR the same filter should work for all record formats, not just CSV.

Engerrs committed:
[#9027] Rebuild implementation for keyset pagination logic

Engerrs committed:
[#9027] Fix lint

Engerrs commented:
Hi @wardi , your last request required full rebuild of the initial implementation, which was a bit challenging to integrate it with the original logic and not break things.

The main updates:
1. Added constant for the operator logic + updated _where_clause to take into account the dict values, which have the operator and value.
2. There is no option of getting the "last_id" in the response, as different types (e.g. xml, json, csv) provide different output of the results, which will require similar parsing as done previously for CSV. Due to this, I added additional query for getting the last_id independently from the chosen type. I checked the performance and there no change in it after the additional query was added.
3. Additionally I had to add "last_id_operator" in the response, as the new approach of using filters with "operator": "value" approach requires the "operator" to make it work. That's why  "last_id_operator" provides this information in the "stream_dump" for the next iteration.
4. Updated the docstring + the .rst file for Datastore. I'm not that good at documentations, so feel free to modify it if needed.

wardi commented on ckanext/datastore/backend/postgres.py:
```diff
@@ -85,6 +85,14 @@
+    'gte': '>=',
+    'lt': '<',
+    'lte': '<=',
+    'in': 'in',
```
let's drop `'in'` because it's not required for pagination and we don't have tests covering its edge cases here

wardi commented on ckanext/datastore/backend/postgres.py:
```diff
@@ -488,6 +499,16 @@ def _where_clauses(
 
 
+def _prepare_where_operator_and_value(value: dict[str, Any]) -> tuple[str, Any]:
+    try:
+        key = next(iter(value))
+        val = value[key]
+    except (StopIteration, KeyError):
+        return '=', value
```
we can ignore the multiple-value dict case too with:
```suggestion
    try:
        [(key, val)] = value.items()
    except ValueError:
        return '=', value
```

Engerrs committed:
[#9027] Minor updates to code

Engerrs commented:
Hi @wardi , done according to your request.

wardi committed:
Merge remote-tracking branch 'origin/master' into update_datastore_dump_download_logic

wardi committed:
get last id in same query, next_page as filter

- not possible for csv/tsv (copy to) methods
- next_page value returned may be added to filter in next call

wardi commented:
I was hoping to fix the issue of duplicated/missing rows that can happen when a rows are added/removed while a dump is running. Getting the next id as part of the same query solves this but for csv/tsv there's still a race condition where the next_id returned won't align with the records generated.

I could solve this by creating a temporary table, but if the temporary space configured in postgres is too small this could result in useless disk writes

I could calculate the `next_id` *first* then modify the query that generates the records to stop at the row with that id instead of using a `limit` but this could mean generating a page that is larger or smaller than what was requested. In the worst case it could even generate an empty set of results before the end of the actual records are released.

It might also be possible to generate csv/tsv data without `copy to` but that would be a much more complicated query (darn you postgres syntax for not being composable)

wardi committed:
[#9027] fix next id with filtered columns

EricSoroos commented:
> I was hoping to fix the issue of duplicated/missing rows that can happen when a rows are added/removed while a dump is running. 

The only reliable way to do this is to rely on the database concurrency management to ensure that we're running in a single transaction or single query. 

> I could solve this by creating a temporary table, but if the temporary space configured in postgres is too small this could result in useless disk writes

I think this would be very bad for dump tables where we're getting hit by multiple concurrent downloads.

wardi committed:
[#9027] adopt filters.py from ckanext-search

wardi committed:
[#9027] fix for unspecified filters

wardi committed:
[#9027] fix for IDatastore custom filters, lint, types

wardi committed:
[#9027] changelog, types

wardi committed:
[#9027] filters docs

wardi committed:
[#9027] test and fix for list or-filters

wardi committed:
[#9027] navl function signatures

wardi committed:
[#9027] move filter docs together

wardi committed:
[#9027] move+edit new search pagination docs

wardi committed:
[#9027] remove _links with next/prev offsets

wardi committed:
[#9027] datastore_search: include_next_page option

wardi committed:
[#9027] docs: upgrade datastore headings to appear in TOC

also update filter and pagination docs

wardi committed:
[#9027] update changelog

wardi committed:
[#9027] test update

wardi committed:
Merge remote-tracking branch 'origin/master' into update_datastore_dump_download_logic

wardi committed:
[#9027] update dump for next_page

wardi commented:
CSV dump time for a real 13M record dataset:

before | after | change
--- | --- | ---
30:09 | 2:08 | 15× faster

Engerrs commented:
@wardi this is great!

wardi committed:
[#9027] dump: fix slow start, from @TomeCirun

wardi committed:
[#9027] test generated pagination filters

wardi committed:
[#9027] dump: check_access instead of empty search

wardi committed:
[#9027] update tests for check_access

wardi committed:
[#9027] datastore_search: alias lookup in auth

move alias lookup into the check_access auth function so that
can call it instead of the action to reliably check permissions

make check_access return the auth funtion result instead of
True so that values can be passed back to actions without
mutating passed dict parameters (in this case the real resource
id is returned when an alias is referenced)

wardi committed:
[#9027] type fixes

wardi committed:
[#9027] missing is not NotAuthorized

wardi committed:
[#9027] dump: copy datastore_search preamble

wardi committed:
Merge remote-tracking branch 'origin/master' into update_datastore_dump_download_logic

EricSoroos commented on ckanext/datastore/blueprint.py:
```diff
@@ -141,11 +148,7 @@ class DictionaryView(MethodView):
-                context={"user": current_user.name, "auth_user_obj": current_user},
-                data_dict={"resource_id": resource_id},
-            )
+            h.check_access("datastore_create", {"resource_id": resource_id})
```
This is incorrect.  Calling h.check_access and not checking the return boolean essentially bypasses the check.

wardi committed:
[#9027] revert check_access change

wardi committed:
[#9027] fix check_access change

EricSoroos commented:
(what this doesn't help is That One Guy downloading an entire 700k element dataset using datastore_search and limit offset pagination, by 100 records at a time)

wardi commented:
@EricSoroos I'm in favor of putting a hard limit on the accepted `offset` values to *encourage* use of more efficient pagination

EricSoroos commented:
I've done that recently on a non-ckan project.

amercader commented:
@wardi it looks like merging this broke master (and 2.12 as I cherry picked the changes):

https://github.com/ckan/ckan/actions/runs/25790338655/job/75754184258

The branch wasn't up to date with master and there are conflicts with https://github.com/ckan/ckan/pull/9298. Basically `WHITELISTED_RESOURCES` is gone from datastore/logic/action.py but is still imported and used from the datastore blueprint. Would you mind updating the blueprint to the new logic against latest master?
